[36기 박승민] Cart 리뷰 요청 반영

pmb087 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)

- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표

- 상품 결제 페이지 구현
- 상품 장바구니 저장기능 구현

<br />

## :: 구현 사항 설명

- **NavCart Component**
> - 메인 페이지에서 상품을 담기 했을경우 장바구니 `DB` 로 해당 상품을 **POST** 합니다.
>
> - **NavBottom Component** 의 자식 컴포넌트인 **NavCart** 컴포넌트는 
부모 컴포넌트의 장바구니 아이콘을 **MouseOver**  하여 렌더링합니다.
>
> -  **NavCart** 가 렌더링 될 때 장바구니 `DB` 를 **GET** 하여 현재 담긴 상품 리스트를 받아옵니다.
>
> - 받아온 데이터로 상품의 세부정보 , 품목 별 수량, 금액 , 총 금액 등을 보여줍니다.
>
> - 데이터로 받아온 상품의 재고 수량을 기준으로 상품 추가 , 감소 버튼을 활성화 , 비활성화 합니다.
>
> - 상품 추가 , 상품 감소 버튼을 누를 때 마다 해당 상품의 구매하고자 하는 수량 (**Quantity**)를 **PATCH** 합니다.
>
> - 각 항목의 휴지통 버튼을 누르면 해당 상품에 관련된 데이터를 장바구니  `DB` 에서 **DELETE** 합니다.
>
> - 구매하기 버튼을 누르면 `**BASE_URL/cart**` 경로로 라우팅됩니다.

- **Cart Component**
> - 페이지 최초 렌더링 시 장바구니 `DB` 를 `GET` 하여 현재 담긴 상품 리스트를 받아옵니다.
>
> - 항목별 기능은 **NavCart Component** 와 동일합니다.
>
> - 결제하기 버튼을 누르면 영수증 모달이 나타납니다.
>
> - 영수증 모달에서는 장바구니 데이터를 근거로하는 총 계산 내역을 알려줍니다.
>
> - 영수증 모달을 닫으면 `DB` 에 구매 요청을 보냅니다. 
>
> - 서버는 장바구니에 있던 상품을 근거로 추천 알고리즘을 업데이트하고, 장바구니 `DB` 를 초기화합니다.
>
> - 해당 요청의 결과로 회원가입시 지급된 크레딧에서 상품 총 결제금액이 제해집니다. 이후 메인 페이지로 되돌아갑니다.

<br />

## :: 성장 포인트 

- 백엔드와 소통하며 **EndPoint_URL** 과 **CRUD API** 문서를 적절히 활용하고 `fetch` 에 활용하는데 익숙해졌습니다.

<br />

## :: 기타 질문 및 특이 사항

Commit:
WIP

Commit:
Merge branch 'main' into feature/cart

Commit:
merge main

Commit:
WIP

Commit:
WIP

Commit:
ADD: Cart Component

ADD: Cart Component
ADD: NavBar Cart Component
장바구니 관련 기능 , 페이지 추가

Commit:
FIX: Refactoring

Commit:
FIX: Remove Console.log

Commit:
Merge branch 'main' into feature/cart

Commit:
FIX: SCSS Nesting

sodalite1204 commented on src/components/Nav/NavBottomContainer/NavBottomContainer.js:
```diff
@@ -20,8 +22,13 @@ function NavBottomContainer({ color }) {
+          </div>
+          <div className="navIconWrap" onMouseOver={() => setIsHover(true)}>
+            <i className="fa-solid fa-basket-shopping navIcon" />
+            {isHover ? <NavCart onMouse={setIsHover} /> : null}
```
굳이 삼항연산자로 true, false 케이스 모두 처리하여 null을 반환할 필요없이 이 부분은 true 케이스만 처리하면 되기 때문에 조건부 렌더링 (`&&` 연산자)을 활용하면 불필요한 코드를 없애 더욱 가독성에 좋아집니다. 
[참고] [공식문서 조건부 렌더링](https://ko.reactjs.org/docs/conditional-rendering.html)

sodalite1204 commented on src/components/Nav/NavBottomContainer/NavCart/NavCart.js:
```diff
@@ -0,0 +1,96 @@
+
+function NavCart({ onMouse }) {
+  const [summaryPrice, setSummaryPrice] = useState();
+  const [productListData, setProductListData] = useState();
```
초기값은 추후에 실질적으로 state에 들어갈 요소의 타입과 일치시켜 넣어주세요. 초기값을 넣어주시는 것 만으로도 어떤 종류의 데이터가 저장되는지 파악할 수 있고, 이것으로도 코드를 파악하는데 많은 도움이 됩니다. 코드의 가독성은 개인 뿐 아니라 결과적으로 팀의 생산성을 높입니다.

sodalite1204 commented on src/components/Nav/NavBottomContainer/NavCart/NavCart.js:
```diff
@@ -0,0 +1,96 @@
+              ? ' 0'
+              : Number(summaryPrice)
+                  .toString()
+                  .replace(/\B(?<!\.\d*)(?=(\d{3})+(?!\d))/g, ',')}
```
많이 쓰이는 부분이면, 공용 함수(utils)로도 만들어서 사용할 수도 있을 것 같네요!
```javascript
const priceToString = price =>
    Number(price)
      .toString()
      .replace(/\B(?<!\.\d*)(?=(\d{3})+(?!\d))/g, ',');

priceToString(3000)
```
참고사항입니다^^

Commit:
FIX: Refactoring , ADD : PaymentModal UserPoint 계산 추가

sodalite1204 commented on src/components/Nav/NavBottomContainer/NavCart/NavCartProduct/NavCartProduct.js:
```diff
@@ -0,0 +1,114 @@
+          <div className="productQuantityControlContainer">
+            <button
+              className="quantityMinus"
+              onClick={() => {
+                setThisQuantity(prev => prev - 1);
+                fetch(`${API.CART}/${orderItemsId}`, {
+                  method: 'PATCH',
+                  headers: {
+                    'Content-Type': 'Application/json',
+                    authorization:
+                      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NiwiZW1haWwiOiJ0ZXN0QDEiLCJpYXQiOjE2NjEzMTg5MDR9.byKbkYPoP3KbJtxPA1txesXuppi3AbJXHqTr2ptmJQc',
+                  },
+                  body: JSON.stringify({
+                    quantity: thisQuantity - 1,
+                  }),
+                });
+                setProductPrice(prev => Number(prev) - Number(price));
+                setSummaryPrice(prev => Number(prev) - Number(price));
+              }}
```
html 안에 여러역할을하는 함수가 있으면, 가독성면에서 떨어질 뿐만아니라, 관심사측면에서도 분리를 해주는 게 좋을 것 같습니다.

sodalite1204 replied:
토큰 부분의 하드 코딩은 추후에 수정할 예정이신가요?

pmb087 replied:
넵 아직 로그인 머지가 안되서 LocalStorage에 토큰이 없는 상태라서요, 로그인 페이지 머지되면 추후 수정하겠습니다~

sodalite1204 commented on src/components/Nav/NavBottomContainer/NavCart/NavCartProduct/NavCartProduct.js:
```diff
@@ -0,0 +1,114 @@
+              }}
+              disabled={thisQuantity === 1}
+            >
+              {' '}
+              -{' '}
```
어떤 코드일까요?

pmb087 replied:
제가 추가한게 아니라.. ESLint 에서 공백을 자동으로 변환해주더라구요 ㅎㅎ;;

sodalite1204 commented on src/components/Nav/NavBottomContainer/NavCart/NavCartProduct/NavCartProduct.js:
```diff
@@ -0,0 +1,114 @@
+            <span className="productQuantity">{thisQuantity}</span>
+            <button
+              className="quantityPlus"
+              onClick={() => {
+                setThisQuantity(prev => prev + 1);
+                fetch(`${API.CART}/${orderItemsId}`, {
+                  method: 'PATCH',
+                  headers: {
+                    'Content-Type': 'Application/json',
+                    authorization:
+                      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NiwiZW1haWwiOiJ0ZXN0QDEiLCJpYXQiOjE2NjEzMTg5MDR9.byKbkYPoP3KbJtxPA1txesXuppi3AbJXHqTr2ptmJQc',
+                  },
+                  body: JSON.stringify({
+                    quantity: thisQuantity + 1,
+                  }),
+                });
+                setProductPrice(prev => Number(prev) + Number(price));
+                setSummaryPrice(prev => Number(prev) + Number(price));
+              }}
```
위 리뷰 참고해주세요!

sodalite1204 commented on src/components/Product/Product.js:
```diff
@@ -33,11 +67,11 @@ function Product({
             <button
               key={color}
-              value={color}
+              value={JSON.stringify({ color: color, colorId: colorId })}
```
JSON.stringify를 사용하셨는데 JSON으로 인코딩이 필요한 이유가 있을까요? 
(아래 부분도)

pmb087 replied:
event.tartget.value 를  사용해서 제가 선택한 color , size 를 매칭시키고싶은었습니다.

Id와 Name을 동시에 취급하려고 하니 객체로 보내야 했는데,
객체와 문자열을 비교하는 형태가 되어서?? 아예 인식을 못하는..

그래서 객체를 문자열로 변경시켜서 비교하기 위해서 JSON.stringify를 사용했습니다.

사실 제가 생각하고 쓴게 아니라 아무리 찾아봐도 저방법밖에 못찾겠더라구요..
더 좋은방법이 있을까요???

sodalite1204 reviewed (commented):
승민님 리뷰 확인해주세요! 공통된 부분은 생략된 부분도 있어서 전체적으로 확인해주세요!

Commit:
FIX: CartSummaryModal

Cart 데이터를 Cart 컴포넌트에서 props로 내려받던 것을
Modal이 렌더링 될 때 fetch로 받아오는 코드로 변경.

CartProduct에서 수량 변경 , 상품 삭제시 DB는 업데이트 되지만 Modal로
넘겨주는 props에서는 갱신되지 않기때문에 수정함.

ghost commented:
✅ Sider has resolved the errors or warnings via the commit 36b814e! See the [analysis result](https://sider.review/gh/repos/525273551/pulls/9) for details.

---

You can [turn off such notifications](https://sider.review/gh/repos/525273551/settings#notifications) if unnecessary.

Commit:
Merge branch 'main' into feature/cart

Commit:
Merge

Commit:
FIX: 리뷰 반영

Commit:
FIX: 중복제거

Commit:
Fix: BackgroundColor Change

Commit:
Fix: NavCart Top Fix

Commit:
FIX: Main OverflowX Fix

Commit:
FIX: 리뷰 반영
